docs/architecture/gas/README.md

```diff
@@ -1,27 +1,125 @@
+the shard gets delayed, which delays block production for the entire blockchain,
+reducing latency and throughput for everybody. If the chunk is really late, the
+block producer will decide to not include the chunk at all and inserts an empty
+chunk. The chunk may be included in the next block.
+
+By now, you probably wonder how we can know the time it takes to execute a
+transaction, given that validators use hardware of their choice. Getting these
+timings right is indeed a difficult problem. Or flipping the problem, assuming
+the timings are already known, then we must implement nearcore such that it
+guarantees to operate within the given time constraints. How we tackle this is
+the topic of this chapter.
+
+If you want to learn more about Gas from a user perspective, 
+[Gas basic concepts](https://docs.near.org/concepts/basics/transactions/gas),
+[Gas advanced concepts](https://docs.near.org/concepts/basics/transactions/gas-advanced),
+and [the runtime fee specification](https://nomicon.io/RuntimeSpec/Fees/) are
+good places to dig deeper.
+
+## Hardware and Timing Assumptions
+
+For timing to make sense at all, we must first define hardware constraints. The
+official hardware requirements for a validator is published on
+[near-nodes.io/validator/hardware](https://near-nodes.io/validator/hardware). It
+may change over time but the main principle is that a moderately configured,
+cloud-hosted virtual machine suffices.
+
+For our gas computation, we assume the minimum required hardware. Then we define
+10¹⁵ gas to be executed in at most 1s. We commonly use 1 Tgas (= 10¹² gas) in
+conversation, which corresponds to 1ms execution time.
+
+Obviously, this definition means that a validator running more powerful hardware
+will execute the transactions faster. That is perfectly okay, as far as the
+protocol is concerned we just need to make sure the chunk is available in time.
+If it is ready even faster, no problem.
+
+Less obviously, this means that even a minimally configured validator is often
+idle. Why is that? Well, the hardware must be prepared to execute chunks that
+are always full. But that is rarely the case, as the gas price increases
+exponentially when chunks are full, which would cause traffic to go back
+eventually.
+
+Futhermore, the hardware has to be ready for transactions of all types,
+including transactions chosen by a malicious actor selecting only the most
+complex transactions. Those transactions can also be unbalanced in what
+bottlenecks they hit. For example, a chunk can be filled with transactions that
+fully utilize the CPU's floating point units. Or they could be using all the
+available disk IO bandwidth.
+
+Because the minimum required hardware needs to meet the timing requirements for
+any of those scenarios, the typical, more balanced case is usually computed
+faster than the gas rule states.
+
+## Transaction Gas Cost Model
+
+A transaction is essentially just a list of actions to be executed on the same
```

Comment:
I wonder how pedantic we should be about transactions vs receipts terminology? 

I think its' better to stick with transaction terminology, as that's what we actually use in our everyday communication, if in fact the actual execution happens when we run a receipt.

Reply:
I think of a transaction as more than just `struct Transaction`. Conceptually a tx is the sum of `struct Transaction` and all produced receipts. And for someone just learning about how the transaction runtime works, I think this line of thinking is more useful than getting into the specifics of "actually we execute receipts not transactions". 

So yeah, I agree to stick to transaction terminology in this paragraph. (But for the part about local and remote receipts, I will keep using the correct terms, e.g. "local receipt" and not "local transaction".  I trust the reader to find [2.5. Transactions And Receipts](https://near.github.io/nearcore/architecture/how/tx_receipts.html) if they get completely confused by this.)